src/include/btree_inline.h

```diff
@@ -2073,45 +2077,74 @@ __wt_btcur_skip_page(
+
     /* Don't skip pages in FLCS trees; deleted records need to read back as 0. */
-    if (S2BT(session)->type == BTREE_COL_FIX)
+    if (btree->type == BTREE_COL_FIX)
         return (0);
 
     /*
      * Determine if all records on the page have been deleted and all the tombstones are visible to
      * our transaction. If so, we can avoid reading the records on the page and move to the next
-     * page. We base this decision on the aggregate stop point added to the page during the last
-     * reconciliation. We can skip this test if the page has been modified since it was reconciled.
-     * We also skip this test on an internal page, as we rely on reconciliation to mark the internal
-     * page dirty. There could be a period of time when the internal page is marked clean but the
-     * leaf page is dirty and has newer data than let on by the internal page's aggregated
-     * information.
+     * page.
      *
-     * We are making these decisions while holding a lock for the page as checkpoint or eviction can
-     * make changes to the data structures (i.e., aggregate timestamps) we are reading. It is okay
-     * if the page is not in memory, or gets evicted before we lock it. In such a case, we can forgo
-     * checking if the page has been modified. So, only do a page modified check if the page was in
-     * memory before locking.
+     * Skip this test on an internal page, as we rely on reconciliation to mark the internal page
+     * dirty. There could be a period of time when the internal page is marked clean but the leaf
+     * page is dirty and has newer data than let on by the internal page's aggregated information.
      */
     if (F_ISSET(ref, WT_REF_FLAG_INTERNAL))
         return (0);
 
+    /*
+     * We are making these decisions while holding a lock for the page as checkpoint or eviction can
+     * make changes to the data structures (i.e., aggregate timestamps) we are reading.
+     */
     WT_REF_LOCK(session, ref, &previous_state);
-    if ((previous_state == WT_REF_DISK || previous_state == WT_REF_DELETED ||
+
+    /*
+     * Check the fast-truncate information, there are 3 cases:
+     *
+     * (1) a fast-truncate page in the WT_REF_DELETED state (known to have ft_info.del information);
+     * (2) an on-disk page in the WT_REF_DISK state (read-only pages where a leaf page was first
+     *     instantiated, then evicted, and then read back in);
+     * (3) an in-memory page in the WT_REF_MEM state (read-only pages that have been instantiated).
+     *
+     * #2 and #3 may have ft_info.del set, but it may be NULL, so we check.
+     */
+    page_del = NULL;
+    if (previous_state == WT_REF_DELETED)
+        page_del = ref->ft_info.del;
+    if (F_ISSET(btree, WT_BTREE_READONLY) &&
+      (previous_state == WT_REF_DISK || previous_state == WT_REF_MEM))
+        page_del = ref->ft_info.del;
+    if (page_del != NULL && __wt_txn_visible(session, page_del->txnid, page_del->timestamp)) {
```

Comment:
It seems odd to me that we would skip pages that are visible to the current session. Is this check the right way around?

Reply:
We are skipping the page where the fast truncate operation is visible to us, so there are no updates on the page that are visible to us.